examples/text/plot_hashing_vs_dict_vectorizer.py

```diff
@@ -3,109 +3,293 @@
+# training set grows. Instead of growing the vectors along with a dictionary,
+# feature hashing builds a vector of pre-defined length by applying a hash
+# function `h` to the features (e.g., tokens), then using the hash values
+# directly as feature indices and updating the resulting vector at those
+# indices. Hashing functions tend to map distinct values to the same hash code
+# (hash collisions). As a result, it is impossible to determine what object
+# generated any particular hash code.
+#
+# Because of the above it is impossible to recover the features and the best
+# approach to estimate the number of unique terms in the original dictionary is
+# to count non-vanishing entries of the encoded features. For such purpose we
+# define the following function:
+
+import numpy as np
+
+
+def n_nonzero_columns(X):
+    """Number of columns with at least one non-zero value in a CSR matrix.
+
+    This is useful to count the number of features columns that are effectively
+    active when using the FeatureHasher.
+    """
+    return len(np.unique(X.nonzero()[1]))
+
+
+# %%
+# The default number of features for the `FeatureHasher` is 2**20. Here we set
+# `n_features = 2**18` to illustrate hash collisions. Notice that a large number
+# of features causes larger coefficient dimensions and then requires more
+# training time.
+#
+# **FeatureHasher on frequency dictionaries**
+
+from sklearn.feature_extraction import FeatureHasher
 
-print("FeatureHasher on frequency dicts")
 t0 = time()
-hasher = FeatureHasher(n_features=n_features)
+hasher = FeatureHasher(n_features=2**18)
 X = hasher.transform(token_freqs(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % n_nonzero_columns(X))
-print()
+dict_count_vectorizers["vectorizer"].append(
+    hasher.__class__.__name__ + "\non freq dicts"
+)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {n_nonzero_columns(X)} unique tokens")
+
+# %%
+# **FeatureHasher on raw tokens**
 
-print("FeatureHasher on raw tokens")
 t0 = time()
-hasher = FeatureHasher(n_features=n_features, input_type="string")
+hasher = FeatureHasher(n_features=2**18, input_type="string")
 X = hasher.transform(tokens(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % n_nonzero_columns(X))
+dict_count_vectorizers["vectorizer"].append(
+    hasher.__class__.__name__ + "\non raw tokens"
+)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {n_nonzero_columns(X)} unique tokens")
+
+# %%
+# The number of unique tokens when using the `FeatureHasher` is lower than those
+# obtained using the `DictVectorizer`. This is due to hash collisions.
```

Comment:
Please move this comment to after the first cell that benchmarks `FeatureHasher` (the one with dict inputs).

Also I think it would be interesting to add a new cell that does the same with `n_features=2**20` to show that it's possible to reduce the number of collisions by increasing the feature space.